client/pkg/transport/listener.go

```diff
@@ -527,8 +528,28 @@ func (info TLSInfo) ServerConfig() (*tls.Config, error) {
 		cfg.ClientCAs = cp
 	}
 
-	// "h2" NextProtos is necessary for enabling HTTP2 for go's HTTP server
-	cfg.NextProtos = []string{"h2"}
+	// If manually defined ciphers are provided and TLS1.3 is not allowed, the
+	// list of ciphers may not allow HTTP2.  Test for this and disable if this is the case.
+	if len(cfg.CipherSuites) > 0 && cfg.MaxVersion > 0 && cfg.MaxVersion < tls.VersionTLS13 {
+
+		// Test for the cases where TLS1.3 is not allowed, and thus http2 may not
+		// be possible when select cipherSuites are defined.
+		var foundHTTP2Cipher bool
+		for _, c := range cfg.CipherSuites {
+			switch c {
+			// If one of these cipher suites are defined, then the h2 protocol is possible.
+			case tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256:
+				foundHTTP2Cipher = true
+				break
+			}
+		}
+		info.DisableHTTP2 = !foundHTTP2Cipher
+	}
+
+	if !info.DisableHTTP2 {
```

Comment:
Don't think it makes sense to allow running etcd without grpc which requires HTTP2. 

grpc is the main etcd API and all the clients implement it. We have alternative API via grpc-gateway, but it just translates request to GRPC API which must be running. User could potentially expose separate endpoint without TLS, but then why they would configure cyphers at all.

Would it be better to just abort etcd startup if user doesn't provide HTTP2 capable cypher?

Reply:
Agree it's an edge situation.  I'm happy with handling the case either via graceful shutdown or continued startup provided a clear log message explains the situation.

Reply:
Ok, that makes sense.  Here is an update where it has reverted to a graceful shutdown (which was the issue causing this PR) and added a flag to disable h2.  The idea of adding a *disabling* option is to keep the invocation consistent with the current configurations already in production.

Reply:
> The idea of adding a disabling option is to keep the invocation consistent with the current configurations already in production.

Not sure I understand. Didn't you present a case that you cannot start etcd if provided cypher suites don't support HTTP2? Fixing this should be our priority. Not sure about case of running etcd without HTTP2 yet.

Reply:
They seem closely tied together-- in that choosing a set of select cipher suites will cause the server not to start, as http2 is enabled by default.  The cases for consideration are:

- HTTPS is not enabled - Cipher setting does not matter, and GRPC/HTTP2 can be enabled or disabled at will

- HTTPS is enabled, but HTTP2 ciphers are not allowed - then GRPC will not run, but the etcd service can still serve client requests (although without the HTTP2 protocol as an option)

- HTTPS is enabled, and HTTP2 ciphers are allowed - then GRPC can run (as HTTP2 is allowed), and any served connection can use HTTP or HTTP2 at will

The hard connection for consideration is GRPC relies on HTTP2.

Reply:
My main concern for now is to avoid overcomplicating the etcd server initialization code. For now let's not  add any new behaviors/features (adding  --disable-http2 is a feature), just improve the user experience of misconfiguring the cyphers. This would be a simple improvement that we could consider backporting.